syntax = "proto3";

package file.v1;

option go_package="file/v1;filev1";

service FileService {
  rpc UploadFile(stream UploadFileRequest) returns (FileResponse);
  rpc CreateFile(CreateFileRequest) returns (FileResponse);
  rpc ListFiles(ListFilesRequest) returns (ListFilesResponse);
  rpc GetFile(GetFileRequest) returns (FileResponse);
  rpc DownloadFile(DownloadFileRequest) returns (stream DownloadFileResponse);
  rpc DeleteFile(DeleteFileRequest) returns (DeleteFileResponse);
  rpc InitiateMultipartUpload(InitiateMultipartUploadRequest) returns (InitiateMultipartUploadResponse);
  rpc UploadPart(UploadPartRequest) returns (UploadPartResponse);
  rpc CompleteMultipartUpload(CompleteMultipartUploadRequest) returns (FileResponse);
  rpc AbortMultipartUpload(AbortMultipartUploadRequest) returns (AbortMultipartUploadResponse);
}

message File {
  string id = 1;
  string user_id = 2;
  string filename = 3;
  int64 size = 4;
  string content_type = 5;
  int64 created_at = 6;
}

message CreateFileRequest {
  string filename = 1;
  int64 size = 2;
  string content_type = 3;
}

message GetFileRequest {
  string id = 1;
}

message ListFilesRequest {}

message ListFilesResponse {
  repeated File files = 1;
}

message FileResponse {
  File file = 1;
}

message DeleteFileRequest {
  string id = 1;
  string user_id = 2;
}

message DeleteFileResponse {
  bool success = 1;
}

message UploadFileRequest {
  oneof data {
    UploadFileMetadata metadata = 1;
    bytes chunk = 2;
  }
}

message UploadFileMetadata {
  string filename = 1;
  string content_type = 2;
}

message DownloadFileRequest {
  string id = 1;
}

message DownloadFileResponse {
  oneof data {
    DownloadFileMetadata metadata = 1;
    bytes chunk = 2;
  }
}

message DownloadFileMetadata {
  string filename = 1;
  string content_type = 2;
  int64 size = 3;
}

message InitiateMultipartUploadRequest {
  string filename = 1;
  string content_type = 2;
  int64 total_size = 3;
}

message InitiateMultipartUploadResponse {
  string upload_id = 1;
  int32 chunk_size = 2;
  int32 total_parts = 3;
}

message UploadPartRequest {
  string upload_id = 1;
  int32 part_number = 2;
  bytes chunk = 3;
}

message UploadPartResponse {
  string etag = 1;
  int32 part_number = 2;
}

message CompleteMultipartUploadRequest {
  string upload_id = 1;
  repeated PartInfo parts = 2;
}

message PartInfo {
  int32 part_number = 1;
  string etag = 2;
}

message AbortMultipartUploadRequest {
  string upload_id = 1;
}

message AbortMultipartUploadResponse {
  bool success = 1;
}